### ISSUE TITLE
How to make anonymous stores, like component-specific?

### ISSUE THREAD

**Issue by akloster** [pmndrs/zustand#398]
My use case is that I have a component with a very complicated state. Normally I would use just useState or maybe useReducer, but this time I thought I give Zustand a try.

Anyway, there could theoretically be more than one instance of the component on a page, so it would be nice to create one store for every instance.

Is there a good way to do this?

The ways I could think of so far don't work. I tried to recreate the useStore hook on every render, but that would mean the store is reinitialized at every render, including the state, which is not what I want.

**Comment by dai-shi** [comment]
Creating a store in component lifecycle is possible, but has a pitfall of violating rules of hooks. So, we don't officially write in docs, and educate it's only for expert users.

You could either use useState lazy initialization or useRef lazy initialization.

```js
import create from 'zustand'

const Component1 = () => {
  const [useStore] = useState(() => create(...))
  const { ... } = useStore()
  ...
}

const Component2 = () => {
  const storeRef = useRef()
  if (!storeRef.current) {
    storeRef.current = create(...)
  }
  const useStore = storeRef.current
  const { ... } = useStore()
  ..
}
```

If you are not sure what they are doing, it's not very recommended to use this component specific usage.

Alternatively, `zustand/context` helps in some similar use cases, and it is designed to avoid the pitfall.

**Comment by charkour** [comment]
This is a pattern I use to create anonymous stores using closures. @dai-shi, I believe this is a third option to create anonymous, component-specific stores and that it also follows the rules of hooks. Is that correct? Here is a [working demo](https://stackblitz.com/edit/react-ts-f3ea7q?file=index.tsx). Thanks

```tsx
// factory to create anonymous stores
const createPersonalStore = (
  createState: StateCreator<TState> | StoreApi<TState>
) => {
  const useStore = create(createState)
  return useStore
}

// hook to use the factory
const usePersonalStore = (
  createState: StateCreator<TState> | StoreApi<TState>
) => {
  const useStore = createPersonalStore(createState)
  return useStore
}

// then use in the component
const Component = () => {
  const useStore = React.useMemo(() => {
    return usePersonalStore(set => ({
      bears: 0,
      increasePopulation: () => set(state => ({ bears: state.bears + 1 })),
      removeAllBears: () => set({ bears: 0 })
    }))
  }, [usePersonalStore])


  const bears = useStore(state => state.bears)
  const increasePopulation = useStore(state => state.increasePopulation)

  return (
    <div>
      <p>{bears}</p>
      <button onClick={increasePopulation}>Increase</button>
    </div>
  )
}
```

**Comment by charkour** [comment]
I think my way is just a complicated version of useState lazy initialization.

**Comment by dai-shi** [comment]
@charkour 
Your pattern is basically this.
```js
  // in component
  const useStore = useMemo(() => create(...), [])
```
Which could be the cleanest code, but React docs says, we can't rely on useMemo as it can throw away the cache.

**Comment by akloster** [comment]
Thank you, I think those answers will help me. Currently I am using a global map to keep the stores, with the key being the url parameter passed into the components. That seems to work, and has additional advantages/disadvantages.

**Comment by dai-shi** [comment]
If you could prepare the stores (in the map) in advance outside React, it would fit well with zustand architecture.

**Comment by n1ru4l** [comment]
You can also use React.useState for initializing the hook/value which is less verbose

https://github.com/dungeon-revealer/dungeon-revealer/blob/16b49db38cdc4d9e7634ec1c02fd11edaa0271a4/src/shared-token-state.tsx#L53

**Comment by ivanyurchenk0** [comment]
I also added a small demo to show a current behavior that is looking very strange in terms of context usage https://codesandbox.io/s/ivanyur4enk0-zustand-createcontext-issue-ym9fp if compare it even with react context. I did not find any details about that behavior on https://github.com/pmndrs/zustand#react-context page that assume single provider and all contexts will work with the same state. Is it possible to extend zustand createContext to support that without using React provider that required default value and all other boilerplate code from upper sample?
